File: leetcode/2/AddTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}
     
var addTwoNumbers = function(l1, l2) {
    let retList = new ListNode(0);
    let curList = retList;
    let carry = 0;
    while(l1 || l2  || carry > 0)
    {
        let sum = (l1 ? l1.val : 0)+ (l2 ? l2.val : 0) + carry;
        
        carry = Math.floor(sum /10);
        
        curList.next = new ListNode(sum%10);
        curList = curList.next;
        
        if(l1)
            l1 = l1.next;
        
        if(l2)
            l2 = l2.next;
    }
    
    return retList.next;
};

let l1 = new ListNode(2);
l1.next = new ListNode(4);
l1.next.next = new ListNode(3);
let l2 = new ListNode(5);
l2.next = new ListNode(6);
l2.next.next = new ListNode(4);
//System.out.println(sl.addTwoNumbers(l1, l2));
let ret = addTwoNumbers(l1, l2);
while(ret != null)
{
    console.log(ret.val);
    ret = ret.next;
} 

File: leetcode/2/AddTwoNumbers.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, _next = nil)
        @val = val
        @next = _next
    end
end
# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
    retNode = ListNode.new(0)
    cur = retNode
    carry = 0
    while l1 || l2  || carry > 0
        sum = (l1 ? l1.val : 0) + (l2 ? l2.val : 0) + carry
        # carry = (sum/10).floor
        carry = sum/10
        cur.next =  ListNode.new(sum%10)
        cur = cur.next

        if l1
            l1 = l1.next
        end
        
        if l2
            l2 = l2.next
        end    
    end
    retNode.next
end


l1 = ListNode.new(2);
l1.next = ListNode.new(4);
l1.next.next = ListNode.new(3);
l2 = ListNode.new(5);
l2.next = ListNode.new(6);
l2.next.next = ListNode.new(4);

ret = add_two_numbers(l1, l2);
while ret
    puts(ret.val);
    ret = ret.next;
end    